invenio_pidstore/providers/datacite.py

```diff
@@ -31,20 +38,71 @@ class DataCiteProvider(BaseProvider):
     default_status = PIDStatus.NEW
     """Default status for newly created PIDs by this provider."""
 
+    doi_prefix_regexp = re.compile(
+        r"10\.\d+(\.\d+)*$"
+    )
+
+    @classmethod
+    def valid_doi_prefix(cls, prefix):
+        """Matches if prefix is a DOI prefix.
+
+        Potential TODO: add to idutils module.
+        """
+        return cls.doi_prefix_regexp.match(prefix)
+
+    @classmethod
+    def generate_doi(cls, prefix, suffix_length=10, split_every=5,
+                     checksum=True):
+        """Generate random DOI with `prefix`."""
+        if not cls.valid_doi_prefix(prefix):
+            logger.exception("Invalid DOI prefix", extra={'prefix': prefix})
+            return None
+
+        if checksum and suffix_length < 3:
+            logger.exception(
+                "Invalid suffix_length. At least 3 if checksum enabled",
+                extra={'suffix_length': suffix_length}
+            )
+            return None
+
+        generator = random.SystemRandom()
+        length = suffix_length - 2 if checksum else suffix_length
+        number = generator.getrandbits(length * 5)
+
+        return (
+            prefix +
+            "/" +
+            base32.encode(
+                number,
+                split_every=split_every,
+                checksum=checksum
+            )
+        )
+
     @classmethod
-    def create(cls, pid_value, **kwargs):
+    def create(cls, pid_value=None, prefix=None, **kwargs):
         """Create a new record identifier.
 
         For more information about parameters,
         see :meth:`invenio_pidstore.providers.BaseProvider.create`.
 
         :param pid_value: Persistent identifier value.
+        :param prefix: DOI prefix if pid_value is None
         :params **kwargs: See
             :meth:`invenio_pidstore.providers.base.BaseProvider.create` extra
             parameters.
         :returns: A :class:`invenio_pidstore.providers.DataCiteProvider`
             instance.
         """
+        if pid_value is None:
```

Comment:
I am wondering if this is a not backward compatible change. I am not sure someone relies of the format of the DOI suffix... So if this it is used in an existing Invenio, then suddenly new DOIs might have different suffix patter if the `create` method is used. What is your opinion?

Reply:
You mean the `else` clause with `normalize_pid(pid_value, 'doi')`, right? That's a fair point. Perhaps the `pid_value` passed was non-normalized for some reason and now it will be normalized... Do we know if that is common? I don't know if it would be breaking things necessarily, but it will do them differently and perhaps in trying to be helpful it turns out not to be. I can remove it for now and get rid of idutils dependency (for now!).